File: 994-rotting-oranges/994-rotting-oranges.java
class Solution {
    public class Pair{
        int row;
        int col;
        
        Pair(int row,int col){
            this.row=row;
            this.col=col;
        }
    }
    
    public int orangesRotting(int[][] arr) {
        LinkedList<Pair>q=new LinkedList<>();
        
        int fresh=0;
        for(int i=0;i<arr.length;i++){
            for(int j=0;j<arr[0].length;j++){
                if(arr[i][j]==1){
                    fresh++;
                }else if(arr[i][j]==2){
                    q.addLast(new Pair(i,j));
                }
            }
        }
        
        if(fresh==0)return 0;
        
        int level=-1;
        int [][]dir={{-1,0},{0,1},{1,0},{0,-1}};
            
        while(q.size()>0){
            int size=q.size();
             level++;
            
            while(size-->0){
                Pair rem=q.removeFirst();
                
                for(int i=0;i<dir.length;i++){
                    int rdash=rem.row+dir[i][0];
                    int cdash=rem.col+dir[i][1];
                    
 if(rdash>=0 && cdash>=0 && rdash<arr.length && cdash<arr[0].length && arr[rdash][cdash]==1){
        q.addLast(new Pair(rdash,cdash));
        arr[rdash][cdash]=0;
        fresh--;
    }
                }
            }
        }
        
        if(fresh!=0){
            return -1;
        }else{
            return level;
        }
        
    }
}

File: 994-rotting-oranges/994-rotting-oranges.cpp
class Solution {
public:
    vector<vector<int>>dir={{1,0},{-1,0},{0,1},{0,-1}};
    
    int orangesRotting(vector<vector<int>>& grid) {
        queue<int>q;
        int freshOranges=0;
        
        int n=grid.size();
        int m=grid[0].size();
        
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j]==1){
                    freshOranges++;
                }else if(grid[i][j]==2){
                    q.push(i*m+j);
                }
            }
        }
        
        if(freshOranges==0)return 0;
        
        int level=0;
        
        while(q.size()!=0){
            int size=q.size();
            
            while(size-->0){
                int val=q.front();q.pop();
                
                int r=val/m;
                int c=val%m;
                
                for(auto &d:dir){
                 
                    int rr=r + d[0];
                    int cc=c + d[1];
                    
                    if(rr>=0 && cc>=0 && rr<n && cc<m && grid[rr][cc]==1){
                        grid[rr][cc]=2;
                        q.push(rr*m+cc);
                        freshOranges--;
                        
                        if(freshOranges==0)return level+1;
                    }
                }
                
            }
            level++;
        }
        return -1;
        
    }
};